File: stats.py
"""
stats.py -- streaming statistics for TVLA.

TVLA (Test Vector Leakage Assessment) compares the timing distributions of two
input classes. If a Welch's two-sample t-test on their per-call cycle counts
produces |t| above a threshold (the community convention is 4.5), the two
distributions differ with overwhelming confidence -- i.e. execution time
depends on which class the input belonged to, which for a secret-dependent
class split means an exploitable timing leak.

Everything is computed online with Welford's algorithm so the live UI can show
the t-value converging without ever holding all samples in memory. Pure Python,
no numpy dependency required.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class RunningStat:
    """Numerically stable online mean/variance (Welford)."""
    n: int = 0
    mean: float = 0.0
    m2: float = 0.0          # sum of squares of deviations from the mean

    def push(self, x: float) -> None:
        self.n += 1
        delta = x - self.mean
        self.mean += delta / self.n
        self.m2 += delta * (x - self.mean)

    @property
    def variance(self) -> float:
        return self.m2 / (self.n - 1) if self.n > 1 else 0.0

    @property
    def std(self) -> float:
        return math.sqrt(self.variance)


@dataclass
class WelchTest:
    """Streaming Welch's t-test between two classes A and B."""
    a: RunningStat = field(default_factory=RunningStat)
    b: RunningStat = field(default_factory=RunningStat)

    def push_a(self, x: float) -> None:
        self.a.push(x)

    def push_b(self, x: float) -> None:
        self.b.push(x)

    @property
    def n(self) -> int:
        return self.a.n + self.b.n

    @property
    def t(self) -> float:
        """Welch's t statistic. 0.0 until both classes have >= 2 samples."""
        if self.a.n < 2 or self.b.n < 2:
            return 0.0
        va, vb = self.a.variance, self.b.variance
        na, nb = self.a.n, self.b.n
        denom = math.sqrt(va / na + vb / nb)
        if denom == 0.0:
            return 0.0
        return (self.a.mean - self.b.mean) / denom

    @property
    def dof(self) -> float:
        """Welch-Satterthwaite degrees of freedom."""
        va, vb = self.a.variance, self.b.variance
        na, nb = self.a.n, self.b.n
        if na < 2 or nb < 2:
            return 0.0
        sa, sb = va / na, vb / nb
        num = (sa + sb) ** 2
        den = (sa ** 2) / (na - 1) + (sb ** 2) / (nb - 1)
        return num / den if den else 0.0

    def confidence_interval(self, z: float = 1.96):
        """Approx CI for the mean difference (large-sample z, default 95%)."""
        va, vb = self.a.variance, self.b.variance
        na, nb = self.a.n, self.b.n
        if na < 2 or nb < 2:
            return (0.0, 0.0)
        se = math.sqrt(va / na + vb / nb)
        diff = self.a.mean - self.b.mean
        return (diff - z * se, diff + z * se)

    def snapshot(self) -> dict:
        lo, hi = self.confidence_interval()
        return {
            "t": self.t,
            "dof": self.dof,
            "n_a": self.a.n,
            "n_b": self.b.n,
            "mean_a": self.a.mean,
            "mean_b": self.b.mean,
            "var_a": self.a.variance,
            "var_b": self.b.variance,
            "diff": self.a.mean - self.b.mean,
            "ci95": (lo, hi),
        }


def crop_threshold(samples: List[int], percentile: float) -> float:
    """Return the value at `percentile` (0..100) for outlier cropping.

    dudect crops the upper tail before testing because a single preemption or
    interrupt can inject a huge outlier that inflates variance and hides real
    leaks. Callers typically drop samples above this threshold.
    """
    if not samples:
        return math.inf
    ordered = sorted(samples)
    idx = min(len(ordered) - 1, int(len(ordered) * percentile / 100.0))
    return float(ordered[idx])


# Two-sided p-value approximation from t and dof, for reporting only. Uses the
# regularized incomplete beta function via a continued fraction (Numerical
# Recipes style) so we avoid a scipy dependency.
def _betacf(a: float, b: float, x: float) -> float:
    MAXIT, EPS, FPMIN = 200, 3e-12, 1e-300
    qab, qap, qam = a + b, a + 1.0, a - 1.0
    c = 1.0
    d = 1.0 - qab * x / qap
    if abs(d) < FPMIN:
        d = FPMIN
    d = 1.0 / d
    h = d
    for m in range(1, MAXIT + 1):
        m2 = 2 * m
        aa = m * (b - m) * x / ((qam + m2) * (a + m2))
        d = 1.0 + aa * d
        if abs(d) < FPMIN:
            d = FPMIN
        c = 1.0 + aa / c
        if abs(c) < FPMIN:
            c = FPMIN
        d = 1.0 / d
        h *= d * c
        aa = -(a + m) * (qab + m) * x / ((a + m2) * (qap + m2))
        d = 1.0 + aa * d
        if abs(d) < FPMIN:
            d = FPMIN
        c = 1.0 + aa / c
        if abs(c) < FPMIN:
            c = FPMIN
        d = 1.0 / d
        delta = d * c
        h *= delta
        if abs(delta - 1.0) < EPS:
            break
    return h


def _betai(a: float, b: float, x: float) -> float:
    if x <= 0.0:
        return 0.0
    if x >= 1.0:
        return 1.0
    lbeta = math.lgamma(a + b) - math.lgamma(a) - math.lgamma(b)
    bt = math.exp(lbeta + a * math.log(x) + b * math.log(1.0 - x))
    if x < (a + 1.0) / (a + b + 2.0):
        return bt * _betacf(a, b, x) / a
    return 1.0 - bt * _betacf(b, a, 1.0 - x) / b


def two_sided_p(t: float, dof: float) -> float:
    """Two-sided p-value for a t statistic with `dof` degrees of freedom."""
    if dof <= 0:
        return 1.0
    return _betai(0.5 * dof, 0.5, dof / (dof + t * t))

File: rust/stats.rs
//! The statistical core of the leakage assessment.
//!
//! TVLA (Goodwill–Jun–Jaffe–Rohatgi, 2011) reduces "does this code leak in the
//! timing channel?" to a two-sample **Welch t-test** between a *fixed* input
//! class (A) and a *random* input class (B). If their cycle-count distributions
//! are distinguishable, the timing is input-dependent.
//!
//! ## The threshold, honestly
//!
//! The conventional decision line is |t| > 4.5. For a **single** Welch test at
//! TVLA sample sizes that corresponds to a two-sided false-positive probability
//! on the order of 1e-5. We keep that interpretation for the *raw* t, which is
//! computed over the whole stream and reported as the primary statistic.
//!
//! Real measurements also have a heavy upper tail (preemptions, interrupts,
//! migrations) that can both *manufacture* a spurious difference and *mask* a
//! real one. Following dudect (Reparaz–Balasch–Gierlichs, 2017) we therefore
//! also compute the t-test after discarding the extreme upper tail, at a
//! **small fixed set** of crops, and report the largest resulting |t| as a
//! `tail_robust_t`.
//!
//! Two deliberate choices distinguish this from a naive "max over many crops":
//!
//!   * The crop set is tiny (3 nested crops), so the multiple-comparison
//!     inflation of the false-positive rate is negligible rather than material.
//!   * We do **not** claim the clean 1e-5 calibration for `tail_robust_t`. It
//!     is a *sensitivity* figure: a leak that only surfaces after cropping is
//!     still a leak, but a near-threshold tail-robust value on a noisy host
//!     warrants re-running pinned and quiesced before you trust it.
//!
//! The confidence interval on the mean difference uses the large-sample normal
//! approximation (z), which is indistinguishable from Student's t at n ≫ 10^4.

use std::collections::VecDeque;

/// Canonical TVLA decision line.
pub const THRESHOLD: f64 = 4.5;

/// z for a two-sided 95% CI (large-sample normal approximation).
const Z95: f64 = 1.959_963_984_540_054;

/// Upper-tail crops for the tail-robust statistic: fraction of samples kept.
/// Small and nested on purpose (see module docs).
const CROPS: [f64; 3] = [1.0, 0.95, 0.90];

/// Welford's online mean/variance — O(1) per sample, flat memory.
#[derive(Clone, Default)]
pub struct Online {
    pub n: u64,
    pub mean: f64,
    m2: f64,
}

impl Online {
    pub fn new() -> Online {
        Online::default()
    }

    #[inline]
    pub fn push(&mut self, x: f64) {
        self.n += 1;
        let delta = x - self.mean;
        self.mean += delta / self.n as f64;
        self.m2 += delta * (x - self.mean);
    }

    /// Sample variance (n − 1 denominator). Zero until two samples.
    pub fn var(&self) -> f64 {
        if self.n > 1 {
            self.m2 / (self.n as f64 - 1.0)
        } else {
            0.0
        }
    }
}

/// Result of a Welch t-test between two classes.
#[derive(Clone, Debug)]
pub struct TResult {
    pub t: f64,
    pub dof: f64,
    pub mean_a: f64,
    pub mean_b: f64,
    pub var_a: f64,
    pub var_b: f64,
    pub n_a: u64,
    pub n_b: u64,
    // ponytail: only read from #[test] assertions, which release builds don't
    // compile — hence the dead_code warning outside `cargo test`.
    #[allow(dead_code)]
    pub diff: f64,
    #[allow(dead_code)]
    pub stderr: f64,
    pub ci_low: f64,
    pub ci_high: f64,
}

impl TResult {
    fn empty() -> TResult {
        TResult {
            t: 0.0,
            dof: 0.0,
            mean_a: 0.0,
            mean_b: 0.0,
            var_a: 0.0,
            var_b: 0.0,
            n_a: 0,
            n_b: 0,
            diff: 0.0,
            stderr: 0.0,
            ci_low: 0.0,
            ci_high: 0.0,
        }
    }

    #[allow(dead_code)]
    pub fn leaks(&self) -> bool {
        self.t.abs() > THRESHOLD
    }
}

/// Welch's unequal-variance t-test between two online accumulators.
pub fn welch(a: &Online, b: &Online) -> TResult {
    let (na, nb) = (a.n, b.n);
    let (va, vb) = (a.var(), b.var());
    if na < 2 || nb < 2 {
        let mut r = TResult::empty();
        r.mean_a = a.mean;
        r.mean_b = b.mean;
        r.var_a = va;
        r.var_b = vb;
        r.n_a = na;
        r.n_b = nb;
        return r;
    }

    let sa = va / na as f64; // = s_a^2 / n_a  (squared standard error, class A)
    let sb = vb / nb as f64;
    let se = (sa + sb).sqrt();
    let diff = a.mean - b.mean;
    let t = if se > 0.0 { diff / se } else { 0.0 };

    // Welch–Satterthwaite effective degrees of freedom.
    let denom = sa * sa / (na as f64 - 1.0) + sb * sb / (nb as f64 - 1.0);
    let dof = if denom > 0.0 { (sa + sb).powi(2) / denom } else { 0.0 };

    let ci = Z95 * se;
    TResult {
        t,
        dof,
        mean_a: a.mean,
        mean_b: b.mean,
        var_a: va,
        var_b: vb,
        n_a: na,
        n_b: nb,
        diff,
        stderr: se,
        ci_low: diff - ci,
        ci_high: diff + ci,
    }
}

/// Bounded, class-labelled **window** of the most recent samples, used only to
/// compute the tail-robust cropped statistic. (This is a recency window, not
/// reservoir sampling — the raw statistic above already covers the full stream
/// with flat memory.)
pub struct TailWindow {
    buf: VecDeque<(u8, f64)>, // (class, value); class 0 = fixed, 1 = random
    cap: usize,
}

impl TailWindow {
    pub fn new(capacity: usize) -> TailWindow {
        TailWindow { buf: VecDeque::with_capacity(capacity), cap: capacity }
    }

    #[inline]
    pub fn push(&mut self, class: u8, value: f64) {
        if self.buf.len() == self.cap {
            self.buf.pop_front();
        }
        self.buf.push_back((class, value));
    }

    #[allow(dead_code)]
    pub fn len(&self) -> usize {
        self.buf.len()
    }

    #[allow(dead_code)]
    /// Visit each recent `(class, value)` sample (used to build the histogram).
    pub fn for_each<F: FnMut(u8, f64)>(&self, mut f: F) {
        for &(cls, v) in self.buf.iter() {
            f(cls, v);
        }
    }

    /// Evaluate Welch's t at each fixed crop and return the (result, kept-fraction)
    /// with the largest |t|. The cutoff is a single pooled percentile applied to
    /// *both* classes, so the comparison is never distorted by cropping the two
    /// classes at different thresholds.
    pub fn evaluate(&self) -> (TResult, f64) {
        if self.buf.len() < 8 {
            return (TResult::empty(), 1.0);
        }
        let mut sorted: Vec<f64> = self.buf.iter().map(|&(_, v)| v).collect();
        sorted.sort_unstable_by(|x, y| x.partial_cmp(y).unwrap());

        let mut best = TResult::empty();
        let mut best_frac = 1.0;
        let mut best_abs = -1.0;

        for &frac in CROPS.iter() {
            let cutoff = percentile(&sorted, frac);
            let mut a = Online::new();
            let mut b = Online::new();
            for &(cls, v) in self.buf.iter() {
                if v <= cutoff {
                    if cls == 0 {
                        a.push(v);
                    } else {
                        b.push(v);
                    }
                }
            }
            let res = welch(&a, &b);
            if res.t.abs() > best_abs {
                best_abs = res.t.abs();
                best = res;
                best_frac = frac;
            }
        }
        (best, best_frac)
    }
}

/// Value at `frac` of a pre-sorted slice (nearest-rank).
fn percentile(sorted: &[f64], frac: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    if frac >= 1.0 {
        return sorted[sorted.len() - 1];
    }
    let idx = ((frac * (sorted.len() - 1) as f64) as usize).min(sorted.len() - 1);
    sorted[idx]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn welford_matches_naive() {
        let data = [10.0, 12.0, 23.0, 23.0, 16.0, 23.0, 21.0, 16.0];
        let mut o = Online::new();
        for &x in &data {
            o.push(x);
        }
        let mean = data.iter().sum::<f64>() / data.len() as f64;
        let var = data.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / (data.len() as f64 - 1.0);
        assert!((o.mean - mean).abs() < 1e-9);
        assert!((o.var() - var).abs() < 1e-9);
    }

    #[test]
    fn identical_classes_do_not_leak() {
        let mut a = Online::new();
        let mut b = Online::new();
        for i in 0..10_000 {
            let x = (i % 100) as f64;
            a.push(x);
            b.push(x);
        }
        assert!(!welch(&a, &b).leaks());
    }

    #[test]
    fn separated_classes_leak() {
        let mut a = Online::new();
        let mut b = Online::new();
        for i in 0..10_000 {
            a.push((i % 100) as f64);
            b.push((i % 100) as f64 + 50.0);
        }
        let r = welch(&a, &b);
        assert!(r.leaks());
        assert!(r.diff < 0.0); // A - B
    }
}